code/game/machinery/autolathe/matterforge.dm

```diff
@@ -0,0 +1,601 @@
+		ui = new(user, src, ui_key, "autolathe.tmpl", capitalize(name), 600, 700)
+
+		// template keys starting with _ are not appended to the UI automatically and have to be called manually
+		ui.add_template("_materials", "autolathe_materials.tmpl")
+		ui.add_template("_reagents", "autolathe_reagents.tmpl")
+		ui.add_template("_designs", "matterforge_designs.tmpl")
+		ui.add_template("_queue", "autolathe_queue.tmpl")
+
+		// when the ui is first opened this is the data it will use
+		ui.set_initial_data(data)
+		// open the new ui window
+		ui.open()
+
+/obj/machinery/matter_nanoforge/attack_hand(mob/user)
+	if(..())
+		return TRUE
+	matter_assoc_list()
+	user.set_machine(src)
+	ui_interact(user)
+	if(!design_list.len)
+		get_designs()
+
+/obj/machinery/matter_nanoforge/Topic(href, href_list)
+	if(..())
+		return
+
+	usr.set_machine(src)
+
+	if(href_list["category"] && categories)
+		var/new_category = text2num(href_list["category"])
+
+		if(new_category && new_category <= length(categories))
+			show_category = categories[new_category]
+		return 1
+
+	if(href_list["eject_material"] && (!current_design || paused || error))
+		var/material = href_list["eject_material"]
+		var/material/M = get_material_by_name(MATERIAL_COMPRESSED_MATTER)
+
+		if(!M.stack_type)
+			return
+
+		var/num = input("Enter sheets number to eject. 0-[stored_material[material]]","Eject",0) as num
+		if(!CanUseTopic(usr))
+			return
+
+		num = min(max(num,0), stored_material[material])
+		eject(num)
+		return 1
+
+
+	if(href_list["add_to_queue"])
+		var/recipe = href_list["add_to_queue"]
+		var/datum/design/picked_design
+
+		for(var/datum/design/f in design_list)
+			var/test = text2path((recipe))
+			if(f.id == test)
+				picked_design = f
+				break
+
+		if(picked_design)
+			var/amount = 1
+
+			if(href_list["several"])
+				amount = input("How many \"[picked_design.id]\" you want to print ?", "Print several") as null|num
+				if(!CanUseTopic(usr) || !(picked_design in design_list))
+					return
+
+			queue_design(picked_design, amount)
+
+		return 1
+
+	if(href_list["remove_from_queue"])
+		var/ind = text2num(href_list["remove_from_queue"])
+		if(ind >= 1 && ind <= queue.len)
+			queue.Cut(ind, ind + 1)
+		return 1
+
+	if(href_list["move_up_queue"])
+		var/ind = text2num(href_list["move_up_queue"])
+		if(ind >= 2 && ind <= queue.len)
+			queue.Swap(ind, ind - 1)
+		return 1
+
+	if(href_list["move_down_queue"])
+		var/ind = text2num(href_list["move_down_queue"])
+		if(ind >= 1 && ind <= queue.len-1)
+			queue.Swap(ind, ind + 1)
+		return 1
+
+
+	if(href_list["abort_print"])
+		abort()
+		return 1
+
+	if(href_list["pause"])
+		paused = !paused
+		return 1
+
+	if(href_list["unfold"])
+		if(unfolded == href_list["unfold"])
+			unfolded = null
+		else
+			unfolded = href_list["unfold"]
+		return 1
+
+/obj/machinery/matter_nanoforge/attackby(obj/item/I, mob/user)
+	if(I.GetComponent(/datum/component/inspiration))
+		GET_COMPONENT_FROM(comp, /datum/component/inspiration, I)
```

Comment:
I wonder if this will work with spacecash as well. And if so, you should put a new variable in the inspiration component that indicates it will not work with this machine in certain cases.

Reply:
what do you mean "new" variable, like one that I manually add?

Reply:
a variable that is applied when initializing the component in the spacecahs.

Reply:
what are these certain cases

Reply:
for example:
![image](https://user-images.githubusercontent.com/49106866/99898155-c56d4300-2c75-11eb-8eee-69003f567ed1.png)

Reply:
I'm working on something similar for an autolathe and what I did was that the perks were inside the inspiration component, so to insert an oddity it must not only have the inspiration component but also an associated perk.

Reply:
> for example:
> ![image](https://user-images.githubusercontent.com/49106866/99898155-c56d4300-2c75-11eb-8eee-69003f567ed1.png)

That does not explain much, all I saw was you repeating yourself when someone asked what it meant.

Reply:
ah okay this is just a suggestion